### PR TITLE
♻️ Low-latency storage permission management

### PR DESCRIPTION
**Pull request by falexwolf**
https://laminlabs.slack.com/archives/C04FPE8V01W/p1706897425702599?thread_ts=1706743094.933629&cid=C04FPE8V01W

### CODE REVIEW

File: lamindb_setup/dev/upath.py [hunk 1]
```diff
@@ -394,43 +391,44 @@ def convert_pathlike(pathlike: Union[str, Path, UPath]) -> UPath:
...
-        # we try to access a public bucket in a different AWS account
-        # or we have the credentials for the wrong AWS acount
-        # or we have expired credentials
-        from ._hub_core import access_aws
-
-        access_aws()
-        # the env variables aren't recognized if previously
-        # another .fs.session._credentials object was already created
+    # test whether AWS credentials are present
+    session = botocore.session.get_session()
```
Comment (Koncopd) [inline lamindb_setup/dev/upath.py:407]: This still happens on every create_path call, but this check only needed once.

File: lamindb_setup/dev/upath.py [hunk 2]
```diff
@@ -394,43 +391,44 @@ def convert_pathlike(pathlike: Union[str, Path, UPath]) -> UPath:
...
+        anon = False
+    # test whether we are on hosted storage or not
+    path_str = path.as_posix()
+    is_hosted_storage = path_str.startswith(hosted_buckets)
+
+    if not is_hosted_storage:
+        # make anon request if no credentials present
+        return UPath(path, cache_regions=True, anon=anon)
+    else:
+        root_folder = path_str.replace("s3://", "").split("/")[0]
```
Comment (Koncopd) [inline lamindb_setup/dev/upath.py:421]: Again this happens on every call of create_path.

File: lamindb_setup/dev/upath.py [hunk 3]
```diff
@@ -394,43 +391,44 @@ def convert_pathlike(pathlike: Union[str, Path, UPath]) -> UPath:
...
+        if root_folder in credentials_cache:
+            credentials = credentials_cache[root_folder]
+        else:
+            from ._hub_core import access_aws
+
+            credentials = access_aws()
+            credentials_cache[root_folder] = credentials
         return UPath(
             path,
```
[the remark this answers is not in the document]
Comment (falexwolf) [reply lamindb_setup/dev/upath.py:430]: Also calling you here!

File: lamindb_setup/dev/upath.py (hunk 2 above)
Comment (falexwolf) [reply lamindb_setup/dev/upath.py:421]: Yes! Calling you!

File: lamindb_setup/dev/upath.py (hunk 1 above)
Comment (falexwolf) [reply lamindb_setup/dev/upath.py:407]: Agreed, we can cache it because we don't use env variables anymore.

**Pull request merged**